src/tools/snapshot/nodes/ObjectSerializationNode.java

```diff
@@ -0,0 +1,236 @@
+    CachedSlotRead[] reads = new CachedSlotRead[definitions.size()];
+
+    // Reads are ordered by the name of the slots
+    // this way the entry order is deterministic and independent of layout changes
+    // same order in replay when reading
+    Collections.sort(definitions, (a, b) -> {
+      return a.getName().getString().compareTo(b.getName().getString());
+    });
+
+    for (int i = 0; i < reads.length; i++) {
+      StorageLocation loc = layout.getStorageLocation(definitions.get(i));
+
+      AbstractDispatchNode next =
+          UninitializedDispatchNode.createLexicallyBound(loc.getSlot().getSourceSection(),
+              loc.getSlot().getName(), clazz.getMixinDefinition().getMixinId());
+
+      reads[i] =
+          loc.getReadNode(SlotAccess.FIELD_READ,
+              DispatchGuard.createSObjectCheck(o),
+              next,
+              false);
+    }
+
+    return reads;
+  }
+
+  protected static final class UninitializedObjectSerializationNode
+      extends ObjectSerializationNode {
+
+    private UninitializedObjectSerializationNode(final SClass clazz) {
+      super(clazz);
+    }
+
+    @Override
+    public void serialize(final Object o, final SnapshotBuffer sb) {
+      if (o instanceof SObject) {
+        clazz.setSerializer(
+            new SObjectSerializationNode(clazz, createReadNodes((SObject) o)));
+        clazz.getSerializer().serialize(o, sb);
+      } else if (o instanceof SObjectWithoutFields) {
+        clazz.setSerializer(
+            new SObjectWithoutFieldsSerializationNode(clazz));
+        clazz.getSerializer().serialize(o, sb);
+      }
+    }
+
+    @Override
+    public Object deserialize(final ByteBuffer sb) {
+      if (clazz.getFactory().hasSlots()) {
+        clazz.setSerializer(
+            new SObjectSerializationNode(clazz, null));
+        return clazz.getSerializer().deserialize(sb);
+      } else {
+        clazz.setSerializer(
+            new SObjectWithoutFieldsSerializationNode(clazz));
+        return clazz.getSerializer().deserialize(sb);
+      }
+    }
+  }
+
+  // These Nodes may need to be replaced in the class when object layouts change
+  protected static final class SObjectSerializationNode
+      extends ObjectSerializationNode {
+
+    private final int                                     fieldCnt;
+    @CompilationFinal @Children private CachedSlotRead[]  fieldReads;
+    @CompilationFinal @Children private CachedSlotWrite[] fieldWrites;
+    private ObjectLayout                                  layout;
+
+    private SObjectSerializationNode(final SClass clazz,
+        final CachedSlotRead[] reads) {
+      super(clazz);
+      layout = clazz.getLayoutForInstancesUnsafe();
+      fieldReads = insert(reads);
+      fieldCnt = fieldReads.length;
+    }
+
+    @ExplodeLoop
+    @Override
+    public void serialize(final Object o, final SnapshotBuffer sb) {
+      assert o instanceof SObject;
+      ((SObject) o).updateLayoutToMatchClass();
```

Comment:
This seems to be something that should be done conditionally, as part of specialization.

Also, this should really be done in a safepoint, see https://github.com/smarr/SOMns/blob/dev/src/som/interpreter/nodes/dispatch/UninitializedDispatchNode.java#L160